### ISSUE TITLE
Using --extend-ignore with line-break warnings

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#676]
In GitLab by @tjlevel12 on Jan 30, 2019, 17:44

I'd like to be able to use the `--extend-ignore` feature that was recently added, but I'm having trouble using this feature while maintaining proper line-break warnings for my projects.

Although I'm asking specifically in regards to the W503/W504 warnings, my question actually extends beyond that. In essence, my question is this: **Is there any way to use the `extend-ignore` feature, while still selectively triggering warnings/errors for the items which exist within the default ignore list?**

It seems handy to automatically track the default ignore list, but as it works now the `extend-ignore` feature seems to be a no-go for any project which needs to use any of the codes in the default ignore list.

I was hoping that I'd be able to `select` W504 as an active warning, but this doesn't seem to be supported: specific codes which exist on both the select and ignore lists are ignored.

# Example

```python
# file.py

# Should trigger a W503 warning
print(1
      + 2)

# Should trigger a W504 warning
print(1 +
      2)
```

```bash
# W503 and W504 are both ignored, as expected
$ flake8 ./file.py

# W503 is ignored as expected, W504 is reported
$ flake8 --ignore=W503 ./file.py
./file.py:8:9: W504 line break after binary operator

# W503 is ignored as expected, W504 is explicitly selected and reported
$ flake8 --select=W504 ./file.py
./file.py:8:9: W504 line break after binary operator

# Neither W503 nor W504 are ignored, as expected
$ flake8 --ignore=E731 ./file.py
./file.py:5:7: W503 line break before binary operator
./file.py:8:9: W504 line break after binary operator

# W503 and W504 are both ignored, as expected
$ flake8 --extend-ignore=E731 ./file.py

# W504 continues to be ignored, even though it's been explicitly selected
$ flake8 --extend-ignore=E731 --select=W504 ./file.py
```

---

flake8 was installed in a virtualenv via `pip install flake8`

Output of `flake8 --bug-report`:

```json
{
  "dependencies": [
    {
      "dependency": "entrypoints",
      "version": "0.3"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.5.2",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.5.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.1.0"
    }
  ],
  "version": "3.7.2"
}
```

**Comment by asottile** [comment]
In GitLab by @tjlevel12 on Jan 30, 2019, 17:55

changed the description

**Comment by asottile** [comment]
In GitLab by @tjlevel12 on Jan 30, 2019, 18:27

I think what this really boils down to is: which takes precedence, a `select` or an `ignore`?

```bash
# W504 is ignored, even though it's been explicitly selected!
$ flake8 --ignore=W504 --select=W504 ./file.py
```

Currently, the codebase turns a "Selected.Explicitly" into a "Decision.Ignored":

```bash
# Get details...
$ flake8 -vv --ignore=W504 --select=W504 ./file.py
<snip>
flake8.style_guide        MainProcess    283 DEBUG    Deciding if "W504" should be reported
flake8.style_guide        MainProcess    284 DEBUG    The user configured "W504" to be "Selected.Explicitly", "Ignored.Explicitly"
flake8.style_guide        MainProcess    284 DEBUG    "W504" will be "Decision.Ignored"
<snip>
```

For me, personally, if I explicitly say that I want to see a specific code, then I'd expect that code to be active, even if I've explicitly ignored it somewhere else.

I think an explicit `select` should override an explicit `ignore` of the same specificity.

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 19:41

It seems the current way that this works is even the default ignore list is considered `Ignored.Explicitly`.  I'd have to scratch my head a bit and think about what changing that to `Ignored.Implicitly` would do (which I *think* would implement the behaviour you're asking for here).

It probably would end up being a behaviour change, but I imagine that it could be desirable to do so anyway?

The other option would be to add a third entry to the matrix and have `Implicitly`, `Default`, `Explicitly` and a whole new set of precedence rules for that, though I'd like to avoid that if possible (as it's difficult enough to reason about the current set of `Ignored` and `Selected` options today)

**Comment by asottile** [comment]
In GitLab by @tjlevel12 on Jan 30, 2019, 20:16

changed the description

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Feb 2, 2019, 06:58

> It seems the current way that this works is even the default ignore list is considered `Ignored.Explicitly`.

I'm pretty sure that should translate to `Ignored.Implicitly` but I haven't looked at the code to be certain.

As for `--ignore W504 --select W504` (from the original comment), I think that I can say confidently that this is junk in causing junk out.

If we wanted to be super specific, we could have a lot more to consider:

- `Implicitly` (defaults)
- `UserConfig` (overrides defaults)
- `ProjectConfig` (overrides UserConfig and Implicitly)
- `CommandLine` (a.k.a., `Explicitly`)

The problem then is that (of course) we don't have the ability to distinguish (now) between different kinds of config so it'd really be


- `Implicitly`
- `ConfigFile`
- `Explicitly`

And that's still a matrix that will be mind-bending to think about given the difficulty that we *already* have in talking about the decision engine. Hell, it was so tricky that I had to pull it apart to make sure I could maintain it and it's still only marginally more understandable than the mess that we had before it.

**Comment by asottile** [comment]
In GitLab by @tjlevel12 on Feb 2, 2019, 12:49

mentioned in commit tjlevel12/flake8@ee8f58adf8480eb02cb57695b9d0214b621637ed

**Comment by asottile** [comment]
In GitLab by @tjlevel12 on Feb 2, 2019, 12:54

mentioned in commit tjlevel12/flake8@ce48997fa01cafd864dcdac9088af070da382bec

**Comment by asottile** [comment]
In GitLab by @gbell12 on Apr 12, 2019, 24:42

Is this related? If not I'll open a new issue:

    $ pycodestyle --select=W504 linebreak.py 
    /home/gbell2/linebreak.py:10:8: W504 line break after binary operator

    $ flake8 --select=W504 linebreak.py 
    <no output>

I understand from the docs that flake8 has its own behavior regarding select/ignore, but I am surprised this simple case isn't doing what I expected.

**Comment by asottile** [comment]
In GitLab by @asottile on Apr 12, 2019, 04:26

@gbell12 I'm unable to reproduce your findings, make sure you're using flake8 3.7.x and don't have any configuration files which explicitly exclude W504.  If you're confused, use `-vvv` to see why it is being turned off:

```console
$ flake8 --version
3.7.7 (mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.6.7 on Linux
$ flake8 t.py
$ flake8 --select=W504 t.py
t.py:2:7: W504 line break after binary operator
```

**Comment by asottile** [comment]
In GitLab by @gbell12 on Apr 13, 2019, 24:56

My mistake - package manager's version of flake8 was very old. Apologies. Working now.

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 16, 2020, 14:46

as above, I believe this is a case of garbage-in-garbage-out? both `--select` and `--ignore` of the same code currently results in ignoring it

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 16, 2020, 14:46

closed
